types/types.go

```diff
@@ -107,12 +107,13 @@ type ImageDestination interface {
 	Reference() ImageReference
 	// FIXME? This should also receive a MIME type if known, to differentiate between schema versions.
 	PutManifest([]byte) error
-	// PutBlob writes contents of stream as a blob identified by digest.
+	// PutBlob writes contents of stream and returns its computed digest and size (both if can be computed).
```

Comment:
The digest and size _must_ be returned, otherwise we won’t be able to build a manifest with the uploaded blob. (And they can of course always be computed locally.)

(Looking at the existing implementations, it seems that all of them will end up doing the `TeeReader` and computing the digest locally; perhaps that might be built as an utility object:
```go
type ReaderForDescriptor interface {
  Reader
  Size() int64 // when Read() returns io.EOF
  Digest() string // when Read() returns io.EOF
}
```
and `PutBlob(stream ReaderForDescriptor, digestIfKnown string) error`.

But that’s a completely optional cleanup.)